File: src/Services/rec_area_server.cpp
#include "ros/ros.h"
//#include "<package_name>/<service_name>.h"
#include "roslearning/rec_area.h"

// Constants
const char *NODE_NAME {"rec_area_server_cpp"};
const char *SERVICE_NAME {"rec_area"};

// Globals
ros::ServiceServer service;

// Handler function for requests on SERVICE_NAME
bool handle_rec_area(roslearning::rec_area::Request& req, roslearning::rec_area::Response& res){
	res.area = req.width * req.height;	// Doing only this is enough
	ROS_INFO("[%s] Request: width = %lf, height = %lf", NODE_NAME, req.width, req.height);
	ROS_INFO("[%s] Sending back response: [%lf]", NODE_NAME, res.area);
	return true;
}
// Service handlers in cpp have two reference parameters as opposed to only one in python

// if not inline, make sure that NodeHandle object is static
inline void setup_node(int argc, char** argv){
	// initialize node
	ros::init(argc, argv, NODE_NAME);
	ros::NodeHandle node;
	service = node.advertiseService(SERVICE_NAME, handle_rec_area);
	// to create ServiceServer object, you need to pass two arguments, service name and handler function address
	ROS_INFO("[%s] Ready to serve", NODE_NAME);
}

int main(int argc, char** argv){
	setup_node(argc, argv);
	ros::spin();
	return 0;
}

File: src/Services/rec_area_server.py
#! /usr/bin/env python

import rospy
from roslearning.srv import rec_area, rec_areaResponse

# CONSTANTS
NODE_NAME = "rec_area_client_py"
SERVICE_NAME = "rec_area"

# GLOBALS
service = None

# request handler
def rec_area_handler(req):
    rospy.loginfo(f"Request: {req.width}, {req.height}")
    rospy.loginfo(f"Returning Response: {req.width * req.height}")
    return rec_areaResponse(req.width * req.height)


# setup node
def setup():
    # initialize node
    rospy.init_node(NODE_NAME)
    # initialize service
    service = rospy.Service(SERVICE_NAME, rec_area, rec_area_handler)
    rospy.loginfo("Ready to serve!")


if __name__ == "__main__":
    setup()
    rospy.spin()    # prevent program from terminating unless killed by user